plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +310,330 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
 	}
 }
 
+static void
+os_upgrade_cancelled_cb (GCancellable *cancellable,
+			 GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	eos_updater_call_cancel (priv->updater_proxy, NULL, NULL, NULL);
+}
+
+static void
+setup_os_upgrade_cancellable (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GCancellable *cancellable = gs_app_get_cancellable (get_os_upgrade (plugin));
+
+	if (g_set_object (&priv->os_upgrade_cancellable, cancellable))
+		g_cancellable_connect (priv->os_upgrade_cancellable,
+				       G_CALLBACK (os_upgrade_cancelled_cb),
+				       plugin, NULL);
+}
+
+static void
+os_upgrade_set_download_by_user (GsApp *app, gboolean value)
+{
+	g_autoptr(GVariant) var = g_variant_new_boolean (value);
+	/* we need to assign it to NULL in order to be able to override it
+	 * (safeguard mechanism in GsApp...) */
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", NULL);
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", var);
+}
+
+static gboolean
+os_upgrade_get_download_by_user (GsApp *app)
+{
+	GVariant *value = gs_app_get_metadata_variant (app, "eos::DownloadByUser");
+	if (value == NULL)
+		return FALSE;
+	return g_variant_get_boolean (value);
+}
+
+static void
+os_upgrade_force_fetch (EosUpdater *updater)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	eos_updater_call_fetch_full (updater,
+				     g_variant_dict_end (&options_dict),
+				     NULL, NULL, NULL);
+}
+
+static gboolean
+os_upgrade_force_fetch_sync (EosUpdater *updater,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	return eos_updater_call_fetch_full_sync (updater,
+						 g_variant_dict_end (&options_dict),
+						 cancellable, error);
+}
+
+static void
+app_ensure_installing_state (GsApp *app)
+{
+	/* ensure the state transition to 'installing' is allowed */
+	if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+}
+
+static gboolean
+eos_updater_error_is_cancelled (const gchar *error_name)
+{
+	return (g_strcmp0 (error_name, "com.endlessm.Updater.Error.Cancelled") != 0);
```

Comment:
This is check for the wrong condition; should be `== 0` to match the cancelled error.

Reply:
Arggh. I had if/else blocks checking for "not cancelled" first, but then changed this last minute to be a one-liner... and screwed up setting the condition.